app/src/main/java/org/coepi/android/ble/covidwatch/BLEAdvertiser.kt

```diff
@@ -0,0 +1,255 @@
+import android.bluetooth.le.AdvertiseData
+import android.bluetooth.le.AdvertiseSettings
+import android.bluetooth.le.BluetoothLeAdvertiser
+import android.content.Context
+import android.os.ParcelUuid
+import android.util.Log
+import org.coepi.android.ble.covidwatch.utils.UUIDs
+import org.coepi.android.ble.covidwatch.utils.toBytes
+import org.coepi.android.ble.covidwatch.utils.toUUID
+import java.util.*
+
+interface BLEAdvertiser {
+    fun startAdvertiser(serviceUUID: UUID?, contactEventUUID: UUID?)
+    fun stopAdvertiser()
+    fun changeContactEventIdentifierInServiceDataField(identifier: UUID)
+}
+
+/**
+ * BLEAdvertiser is responsible for advertising the bluetooth services.
+ * Only one instance of this class is to be constructed, but its not enforced. (for now)
+ * You have been warned!
+ */
+class BLEAdvertiserImpl(private val context: Context, adapter: BluetoothAdapter)
+    : BLEAdvertiser {
+
+    // BLE
+    private val advertiser: BluetoothLeAdvertiser = adapter.bluetoothLeAdvertiser
+    private var bluetoothGattServer: BluetoothGattServer? = null
+    private var advertisedContactEventUUID: UUID? = null
+
+    // CONSTANTS
+    companion object {
+        private const val TAG = "BLEAdvertiser"
+    }
+
+    /**
+     * Callback when advertisements start and stops
+     */
+    private val advertisingCallback: AdvertiseCallback = object : AdvertiseCallback() {
+
+        override fun onStartSuccess(settingsInEffect: AdvertiseSettings) {
+            Log.w(TAG, "onStartSuccess settingsInEffect=$settingsInEffect")
+            super.onStartSuccess(settingsInEffect)
+        }
+
+        override fun onStartFailure(errorCode: Int) {
+            Log.e(TAG, "onStartFailure errorCode=$errorCode")
+            super.onStartFailure(errorCode)
+        }
+    }
+
+    private var bluetoothGattServerCallback: BluetoothGattServerCallback =
+        object : BluetoothGattServerCallback() {
+
+            override fun onServiceAdded(status: Int, service: BluetoothGattService?) {
+                super.onServiceAdded(status, service)
+                Log.i(TAG, "onServiceAdded status=$status service=$service")
+            }
+
+            override fun onCharacteristicReadRequest(
+                device: BluetoothDevice?,
+                requestId: Int,
+                offset: Int,
+                characteristic: BluetoothGattCharacteristic?
+            ) {
+                super.onCharacteristicReadRequest(device, requestId, offset, characteristic)
+
+                var result = BluetoothGatt.GATT_SUCCESS
+                var value: ByteArray? = null
+
+                try {
+                    if (characteristic?.uuid == UUIDs.CONTACT_EVENT_IDENTIFIER_CHARACTERISTIC) {
+                        if (offset != 0) {
+                            result = BluetoothGatt.GATT_INVALID_OFFSET
+                            return
+                        }
+
+                        val newContactEventIdentifier = UUID.randomUUID()
```

Comment:
Hopefully we'll have CenManager call here

Reply:
I deleted `onCharacteristicReadRequest`. The covidwatch team said we will not use the characteristic, only the advertisement data.